src/Modules/SimplCommerce.Module.Inventory/Controllers/WarehouseApiController.cs

```diff
@@ -29,5 +35,174 @@ public WarehouseApiController(IRepository<Warehouse> warehouseRepository)
+            {
+                var address = new Address
+                {
+                    ContactName = model.ContactName,
+                    AddressLine1 = model.AddressLine1,
+                    AddressLine2 = model.AddressLine2,
+                    Phone = model.Phone,
+                    StateOrProvinceId = model.StateOrProvinceId,
+                    CountryId = model.CountryId,
+                    City = model.City,
+                    DistrictId = model.DistrictId,
+                    PostalCode = model.PostalCode
+                };
+
+                var warehouse = new Warehouse
+                {
+                    Name = model.Name,
+                    Address = address
+                };
+
+                using (var transaction = _warehouseRepository.BeginTransaction())
+                {
+                    _addressRepository.Add(address);
+                    await _addressRepository.SaveChangesAsync();
+
+                    _warehouseRepository.Add(warehouse);
+                    await _warehouseRepository.SaveChangesAsync();
+
+                    transaction.Commit();
+                }
+
+                return CreatedAtAction(nameof(Get), new { id = warehouse.Id }, null);
+            }
+            return BadRequest(ModelState);
+        }
+
+        [HttpPut("{id}")]
+        public async Task<IActionResult> Put(long id, [FromBody] WarehouseVm model)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            var warehouse = await _warehouseRepository.Query().FirstOrDefaultAsync(x => x.Id == id);
+            if (warehouse == null)
+            {
+                return NotFound();
+            }
+
+            var address = await _addressRepository.Query().FirstOrDefaultAsync(a => a.Id == model.AddressId);
+
+            if (address != null)
+            {
+                address.ContactName = model.ContactName;
+                address.Phone = model.Phone;
+                address.PostalCode = model.PostalCode;
+                address.StateOrProvinceId = model.StateOrProvinceId;
+                address.CountryId = model.CountryId;
+                address.DistrictId = model.DistrictId;
+
+                await _addressRepository.SaveChangesAsync();
+            }
+
+            warehouse.Address = address;
+            warehouse.Name = model.Name;
+
+            await _warehouseRepository.SaveChangesAsync();
+
+            return Accepted();
+        }
+
+        [HttpDelete("{id}")]
+        public async Task<IActionResult> Delete(long id)
+        {
+            var warehouse = await _warehouseRepository.Query().Include(w=> w.Address).FirstOrDefaultAsync(x => x.Id == id);
+            if (warehouse == null)
+            {
+                return NotFound();
+            }
+
+            try
+            {
+                using (var tranaction = _warehouseRepository.BeginTransaction())
```

Comment:
just call SaveChangesAsync one with out explicitly wrap to a transaction is ok. EF can detected all the changed in the models (in this case removed) the wrap to a single transaction when SaveChanges is called